Add microscopy tutorial (#1036)
ivadomed/ivadomed: docs/source/tutorials/two_class_microscopy_seg_2d_unet.rst

the ``ivadomed/config/`` and the documentation is available in :doc:`../configuration_file`.

In this tutorial we will use the configuration file: ``ivadomed/config/config_microscopy.json``.
First off, copy this configuration file in your local directory (to avoid modifying the source file):

.. code-block:: bash

   cp <PATH_TO_IVADOMED>/ivadomed/config/config_microscopy.json .

Then, open it with a text editor. Below we will discuss some of the key parameters to perform a two-class 2D
microscopy segmentation training.

- ``command``: Action to perform. Here, we want to train a model, so we set the fields as follows:

  .. code-block:: xml

     "command": "train"

- ``path_output``: Folder name that will contain the output files (e.g., trained model, predictions, results).

  .. code-block:: xml

     "path_output": "log_microscopy_sem"

- ``loader_parameters:path_data``: Location of the dataset. As discussed in `Data <../data.html>`__, the dataset
  should conform to the BIDS standard. Modify the path so it points to the location of the downloaded dataset.

  .. code-block:: xml

>      "path_data": "data_axondeepseg_sem"

Author: Good catch! Resolved in e075dee.

- ``loader_parameters:extensions``: File extensions of the microscopy raw data

  .. code-block:: xml

     "extensions": [".png"]

- ``loader_parameters:contrast_params``: Contrasts of interest. In our case, we are training a single contrast model
  with contrast ``SEM``.

  .. code-block:: xml

     "contrast_params": {
         "training_validation": ["SEM"],
         "testing": ["SEM"],
         "balance": {}
     }

- ``loader_parameters:slice_axis``: Orientation of the 2D slice to use with the model.
  2D PNG files must use default ``axial``.

  .. code-block:: xml

     "slice_axis": "axial"

- ``split_dataset:split_method``: Describe the metadata used to split the train/validation/test sets.
  Here, ``sample_id`` from the ``samples.tsv`` file will shuffle all samples, then split them between
  train/validation/test sets.
- ``split_dataset:train_fraction``: Fraction of the dataset's ``sample_id`` in the train set. In our case ``0.6``.
- ``split_dataset:test_fraction``: Fraction of the dataset's ``sample_id`` in the test set. In our case ``0.1``.

  .. code-block:: xml

      "split_method" : "sample_id"
      "train_fraction": 0.6
      "test_fraction": 0.1

> - ``training_time:num_epochs``: The maximum number of epochs that will be run during training. Each epoch is composed

Reviewer: ```suggestion
- ``training_parameters:training_time:num_epochs``: The maximum number of epochs that will be run during training. Each epoch is composed
```

>   of a training part and an evaluation part. It should be a strictly positive integer. In our case, we will use

Reviewer: This is just a suggestion as to me evaluation sounds more on the test set and validation on the validation set, but please ignore if you disagree. Later in the tutorial, there is a section named `Evaluate model` which refers to inference on the testing set. Using evaluation here might be confusing.

```suggestion
  of a training part and a validation part. It should be a strictly positive integer. In our case, we will use
```

Author: Agreed!

  50 epochs.

  .. code-block:: xml

>      "num_epochs": 50

Reviewer: After ~15 epochs I reach ~85% of Dice score on the validation set. We could consider reducing the number of epochs as it's only for a tutorial.

Author: Interesting!
I'll do a training with only 15-20 epochs to see if the segmentation look similar to the segmentations after 50 epochs (I had some surprising results in the past even with similar Dice).
If I can, I'll reduce the number of epochs and update the figure. Thanks!

Author: As I was suspecting, even with a similar Dice after 20 epochs, the segmentation results are not as good (a lot of noise and false positive in the background). I'm not sure why this happens... maybe because the reported Dice after each epochs is a mean of all classes?
Anyway, for now I would prefer to keep the 50 epochs so we have good segmentations like in the figure.

Reviewer: Sounds good to me :)!

- ``default_model:length_2D``: Size of the 2D patches used as model’s input tensors. We recommend using patches
  between 256x256 and 512x512. In our case, we use patches of 256x256.
- ``default_model:stride_2D``: Pixels’ shift over the input matrix to create 2D patches. In our case, we use
  a stride of 244 pixels in both dimensions, resulting in an overlap of 12 pixels between patches.

  .. code-block:: xml

     "length_2D": [256, 256]
     "stride_2D": [244, 244]

> - ``postprocessing:binarize_maxpooling``: Used to binarize predictions across all classes in multiclass models.

Reviewer: ```suggestion
- ``postprocessing:binarize_maxpooling``: Used to binarize predictions across all classes in multiclass models. For each pixel, the class, including the background class, with the highest output probability will be segmented.
```

  .. code-block:: xml

      "binarize_maxpooling": {}

- ``transformation:Resample``: Used to resample images to a common resolution (in mm) before splitting into patches,
  according to each image real pixel size. In our case, we resample the images to a common resolution of 0.0001 mm

>   (0.1 um) in both dimensions.

Reviewer: ```suggestion
  (0.1 μm) in both dimensions.
```

  .. code-block:: xml

     "Resample":
        {
            "hspace": 0.0001,
            "wspace": 0.0001
        },

Train model
-----------

Once the configuration file is ready, run the training:

.. code-block:: bash

   ivadomed -c config_microscopy.json

Alternatively, the "command", "path_output", and "path_data" arguments can be passed as CLI flags
in which case they supersede the configration file, see `Usage <../usage.html>`__.

.. code-block:: bash

   ivadomed --train -c config_microscopy.json --path-data path/to/bids/data --path-output path/to/output/directory

.. note::

   If a `compatible GPU <https://pytorch.org/get-started/locally/>`_ is available, it will be used by default.
   Otherwise, training will use the CPU, which will take a prohibitively long computational time (several hours).

The main parameters of the training scheme and model will be displayed on the terminal, followed by the loss value
on training and validation sets at every epoch. To know more about the meaning of each parameter, go to
:doc:`../configuration_file`. The value of the loss should decrease during the training.

> .. code-block:: console

Reviewer: We should probably update all our tutorials eventually (not the most important thing), but the output in the console doesn't match the output in our tutorials (regardless of the INFO part, some prints don't match): 
![image](https://user-images.githubusercontent.com/49137243/148982559-231097dc-d6eb-44c9-a64f-236114e1f3fb.png)

Author: It may be a mistake on my part when removing the "date/info" part of the output. I'll double-check.

Author: I did not find significant differences but made some minor aesthetic changes in 69b2901.

   RandomAffine: {'degrees': 2.5, 'scale': [0.05, 0.05], 'translate': [0.015, 0.015], 'applied_to': ['im', 'gt']}
   ElasticTransform: {'alpha_range': [100.0, 150.0], 'sigma_range': [4.0, 5.0], 'p': 0.5, 'applied_to': ['im', 'gt']}
   NormalizeInstance: {'applied_to': ['im']}
   Selected transformations for the ['validation'] dataset:
   Resample: {'hspace': 0.0001, 'wspace': 0.0001}
   NormalizeInstance: {'applied_to': ['im']}

   Loading dataset: 100%████████████████████████████████████████████████████████████████ 3/3 [00:00<00:00, 738.48it/s]
   Loaded 76 axial patches of shape [256, 256] for the validation set.
   Loading dataset: 100%████████████████████████████████████████████████████████████████ 6/6 [00:00<00:00, 829.21it/s]
   Loaded 252 axial patches of shape [256, 256] for the training set.
   Creating model directory: log_microscopy_sem/model_seg_rat_axon-myelin_sem

   Initialising model's weights from scratch.

   Scheduler parameters: {'name': 'CosineAnnealingLR', 'base_lr': 1e-05, 'max_lr': 0.01}

   Selected Loss: DiceLoss
   with the parameters: []
   Epoch 1 training loss: -0.6894.
   Epoch 1 validation loss: -0.7908.

> After 50 epochs (see ``"num_epochs"`` in the configuration file), the Dice score on the validation set should be ~85%.

Reviewer: Same comment as above, after 15 epochs I reach this performance.
